[SPARK-23736][SQL] Extending the concat function to support array columns

mn-mikke opened the pull request:
## What changes were proposed in this pull request?
The PR adds a logic for easy concatenation of multiple array columns and covers:
- Concat expression has been extended to support array columns
- A Python wrapper

## How was this patch tested?
New tests added into:
- CollectionExpressionsSuite
- DataFrameFunctionsSuite
- typeCoercion/native/concat.sql

## Codegen examples
### Primitive-type elements
```
val df = Seq(
  (Seq(1 ,2), Seq(3, 4)),
  (Seq(1, 2, 3), null)
).toDF("a", "b")
df.filter('a.isNotNull).select(concat('a, 'b)).debugCodegen()
```
Result:
```
/* 033 */         boolean inputadapter_isNull = inputadapter_row.isNullAt(0);
/* 034 */         ArrayData inputadapter_value = inputadapter_isNull ?
/* 035 */         null : (inputadapter_row.getArray(0));
/* 036 */
/* 037 */         if (!(!inputadapter_isNull)) continue;
/* 038 */
/* 039 */         ((org.apache.spark.sql.execution.metric.SQLMetric) references[0] /* numOutputRows */).add(1);
/* 040 */
/* 041 */         ArrayData[] project_args = new ArrayData[2];
/* 042 */
/* 043 */         if (!false) {
/* 044 */           project_args[0] = inputadapter_value;
/* 045 */         }
/* 046 */
/* 047 */         boolean inputadapter_isNull1 = inputadapter_row.isNullAt(1);
/* 048 */         ArrayData inputadapter_value1 = inputadapter_isNull1 ?
/* 049 */         null : (inputadapter_row.getArray(1));
/* 050 */         if (!inputadapter_isNull1) {
/* 051 */           project_args[1] = inputadapter_value1;
/* 052 */         }
/* 053 */
/* 054 */         ArrayData project_value = new Object() {
/* 055 */           public ArrayData concat(ArrayData[] args) {
/* 056 */             for (int z = 0; z < 2; z++) {
/* 057 */               if (args[z] == null) return null;
/* 058 */             }
/* 059 */
/* 060 */             long project_numElements = 0L;
/* 061 */             for (int z = 0; z < 2; z++) {
/* 062 */               project_numElements += args[z].numElements();
/* 063 */             }
/* 064 */             if (project_numElements > 2147483632) {
/* 065 */               throw new RuntimeException("Unsuccessful try to concat arrays with " + project_numElements +
/* 066 */                 " elements due to exceeding the array size limit 2147483632.");
/* 067 */             }
/* 068 */
/* 069 */             long project_size = UnsafeArrayData.calculateSizeOfUnderlyingByteArray(
/* 070 */               project_numElements,
/* 071 */               4);
/* 072 */             if (project_size > 2147483632) {
/* 073 */               throw new RuntimeException("Unsuccessful try to concat arrays with " + project_size +
/* 074 */                 " bytes of data due to exceeding the limit 2147483632 bytes" +
/* 075 */                 " for UnsafeArrayData.");
/* 076 */             }
/* 077 */
/* 078 */             byte[] project_array = new byte[(int)project_size];
/* 079 */             UnsafeArrayData project_arrayData = new UnsafeArrayData();
/* 080 */             Platform.putLong(project_array, 16, project_numElements);
/* 081 */             project_arrayData.pointTo(project_array, 16, (int)project_size);
/* 082 */             int project_counter = 0;
/* 083 */             for (int y = 0; y < 2; y++) {
/* 084 */               for (int z = 0; z < args[y].numElements(); z++) {
/* 085 */                 if (args[y].isNullAt(z)) {
/* 086 */                   project_arrayData.setNullAt(project_counter);
/* 087 */                 } else {
/* 088 */                   project_arrayData.setInt(
/* 089 */                     project_counter,
/* 090 */                     args[y].getInt(z)
/* 091 */                   );
/* 092 */                 }
/* 093 */                 project_counter++;
/* 094 */               }
/* 095 */             }
/* 096 */             return project_arrayData;
/* 097 */           }
/* 098 */         }.concat(project_args);
/* 099 */         boolean project_isNull = project_value == null;
```

### Non-primitive-type elements
```
val df = Seq(
  (Seq("aa" ,"bb"), Seq("ccc", "ddd")),
  (Seq("x", "y"), null)
).toDF("a", "b")
df.filter('a.isNotNull).select(concat('a, 'b)).debugCodegen()
```
Result:
```
/* 033 */         boolean inputadapter_isNull = inputadapter_row.isNullAt(0);
/* 034 */         ArrayData inputadapter_value = inputadapter_isNull ?
/* 035 */         null : (inputadapter_row.getArray(0));
/* 036 */
/* 037 */         if (!(!inputadapter_isNull)) continue;
/* 038 */
/* 039 */         ((org.apache.spark.sql.execution.metric.SQLMetric) references[0] /* numOutputRows */).add(1);
/* 040 */
/* 041 */         ArrayData[] project_args = new ArrayData[2];
/* 042 */
/* 043 */         if (!false) {
/* 044 */           project_args[0] = inputadapter_value;
/* 045 */         }
/* 046 */
/* 047 */         boolean inputadapter_isNull1 = inputadapter_row.isNullAt(1);
/* 048 */         ArrayData inputadapter_value1 = inputadapter_isNull1 ?
/* 049 */         null : (inputadapter_row.getArray(1));
/* 050 */         if (!inputadapter_isNull1) {
/* 051 */           project_args[1] = inputadapter_value1;
/* 052 */         }
/* 053 */
/* 054 */         ArrayData project_value = new Object() {
/* 055 */           public ArrayData concat(ArrayData[] args) {
/* 056 */             for (int z = 0; z < 2; z++) {
/* 057 */               if (args[z] == null) return null;
/* 058 */             }
/* 059 */
/* 060 */             long project_numElements = 0L;
/* 061 */             for (int z = 0; z < 2; z++) {
/* 062 */               project_numElements += args[z].numElements();
/* 063 */             }
/* 064 */             if (project_numElements > 2147483632) {
/* 065 */               throw new RuntimeException("Unsuccessful try to concat arrays with " + project_numElements +
/* 066 */                 " elements due to exceeding the array size limit 2147483632.");
/* 067 */             }
/* 068 */
/* 069 */             Object[] project_arrayObjects = new Object[(int)project_numElements];
/* 070 */             int project_counter = 0;
/* 071 */             for (int y = 0; y < 2; y++) {
/* 072 */               for (int z = 0; z < args[y].numElements(); z++) {
/* 073 */                 project_arrayObjects[project_counter] = args[y].getUTF8String(z);
/* 074 */                 project_counter++;
/* 075 */               }
/* 076 */             }
/* 077 */             return new org.apache.spark.sql.catalyst.util.GenericArrayData(project_arrayObjects);
/* 078 */           }
/* 079 */         }.concat(project_args);
/* 080 */         boolean project_isNull = project_value == null;
```

Commit:
[SPARK-23736][SQL] Implementation of the concat_arrays function concatenating multiple array columns into one.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -408,6 +408,7 @@ object FunctionRegistry {
     expression[MapValues]("map_values"),
     expression[Size]("size"),
     expression[SortArray]("sort_array"),
+    expression[ConcatArrays]("concat_arrays"),
```
Why not reusing `concat`?

concat(array1, array2, ..., arrayN) -> array ?

mn-mikke replied:
I've already played with this option in my mind, but I'm not sure how concat would be categorized.

Currently, concat is defined as a pure string operation:
  /**
   * @group string_funcs
   * @since 1.5.0
   */
  @scala.annotation.varargs
  def concat(exprs: Column*): Column

Whereas the functionality in this PR belongs rather to the collection_funcs group.

Having just one function for both expressions would be elegant, but can you advise what group should be assigned to concat?

gatorsmile replied:
How about move it to collection functions?

maropu commented:
Thinks for this work! One question; why do you think we need to support this api in Spark native? Other libraries support this as first-class?

mn-mikke commented:
@maropu What other libraries do you mean? I'm not aware of any library providing this functionality on top Spark SQL.

When using Spark SQL as an ETL tool for structured and nested data, people are forced to use UDFs for transforming arrays since current api for array columns is lacking. This approach brings several drawbacks:
- bad code readability
- Catalyst is blind when performing optimizations
- impossibility to track data lineage of the transformation (a key aspect for the financial industry, see [Spline](https://absaoss.github.io/spline/) and [Spline paper](https://github.com/AbsaOSS/spline/releases/download/release%2F0.2.7/Spline_paper_IEEE_2018.pdf))

So my colleagues and I decided to extend the current Spark SQL API with well-known collection functions like concat, flatten, zipWithIndex, etc. We don't want to keep this functionality just in our fork of Spark, but would like to share it with others.

WeichenXu123 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -699,3 +699,88 @@ abstract class TernaryExpression extends Expression {
+   * @param f a function that accepts a sequence of non-null evaluation result names of children
+   *          and returns Java code to compute the output.
+   */
+  protected def nullSafeCodeGen(
```
This method looks almost the same with the one in `BinaryExpression`.  Can you avoid the code duplication ?

gatorsmile replied:
We will combine it with `concat`

mn-mikke replied:
@WeichenXu123 I do agree that there are strong similarities in the code.

If you take a look at `UniryExpression`, `BinaryExpression`, `TernaryExpression`, you will see that methods responsible for null save evaluation and code generation are the same except the number of parameters. My intention has been to generalize the methods into the `NullSaveEvaluation` trait and remove the original methods in a different PR once the trait is in. I didn't want to create a big bang PR because of one additional function in API.

maropu replied:
This is a refactoring issue, so I think we should discuss in a separate ticket. Can you make this pr minimal as much as possible?

mn-mikke replied:
Ok, will try.

gatorsmile commented:
@maropu Maybe you can help @mn-mikke  review this PR ?  Will open an umbrella JIRA for the built-in functions we plan to do in Apache 2.4. In the list, we have multiple for operating nested data.

maropu commented:
ok, I'll check later!

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -3046,6 +3046,14 @@ object functions {
+   * @group collection_funcs
+   * @since 2.4.0
+   */
+  def concat_arrays(columns: Column*): Column = withExpr { ConcatArrays(columns.map(_.expr)) }
```
We need to add this func. in `sql/functions` here? It seems we might recommend users to use these kinds of functions via `selectExpr`, so is it okay to add this only in `FunctionRegistry`? Thoughts? @viirya @gatorsmile

maropu commented:
The current code can't handle inner arrays;
```
scala> sql("select concat_arrays(array(1, 2, array(3, 4)), array(5, 6, 7, 8))").show
org.apache.spark.sql.AnalysisException: cannot resolve 'array(1, 2, array(3, 4))' due to data type mismatch: input to function array should all be the same type, but it's [int, int, array<int>]; line 1 pos 21;
'Project [unresolvedalias('concat_arrays(array(1, 2, array(3, 4)), array(5, 6, 7, 8)), None)]
+- OneRowRelation
```

IMHO, it's better to make this function behaviour the same with postgresql: https://www.postgresql.org/docs/10/static/functions-array.html
Could you brush up code to handle this?

maropu commented:
We should handle different (and compatible) typed arrays in this funs?
```
scala> sql("select concat_arrays(array(1L, 2L), array(3, 4))").show
org.apache.spark.sql.AnalysisException: cannot resolve 'concat_arrays(array(1L, 2L), array(3, 4))' due to data type mismatch: input to function concat_arrays sh
ould all be the same type, but it's [array<bigint>, array<int>]; line 1 pos 7;
'Project [unresolvedalias(concat_arrays(array(1, 2), array(3, 4)), None)]
+- OneRowRelation
```
Also, could you add more tests for this case in `SQLQueryTestSuite`? probably, we can add a new test file like `concat_arrays.sql` in `typeCoercion.native`.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +289,152 @@ case class ArrayContains(left: Expression, right: Expression)
+      > SELECT _FUNC_(array(1, 2, 3), array(4, 5), array(6));
+       [1,2,3,4,5,6]
+  """)
+case class ConcatArrays(children: Seq[Expression]) extends Expression with NullSafeEvaluation {
```
Can we add a common base class (e.g., `ConcatLike`) for handling nested `ConcatArrays` in the optimizer(`CombineConcat`)?

https://github.com/apache/spark/blob/e4bec7cb88b9ee63f8497e3f9e0ab0bfa5d5a77c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala#L649

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +289,152 @@ case class ArrayContains(left: Expression, right: Expression)
+      |Object[] $arrayName = new Object[$numElemName];
+      |int $counter = 0;
+      |$assignments
+      |$arrayDataName = new $genericArrayClass($arrayName);
```
+1, can we reuse the `UnsafeArrayWriter` logic for this case?

mn-mikke replied:
Really like this idea! I think it would require moving the complex type insertion logic from `InterprettedUnsafeProjection` directly to `UnsafeDataWriter` and introduce in that way write methods for complex type fields. I'm not sure whether this big refactoring task is still in the scope of this PR.

Also see that we could improve codeGen of `CreateArray` in the same way.

maropu replied:
You couldn't use `UnsafeArrayData` in the complex case?

mn-mikke replied:
Yeah, currently there are no `write` methods on `UnsafeArrayWriter` or `set` methods on `UnsafeArrayData` that we could leverage for complex types. In theory, we could follow the same approach as in `InterprettedUnsafeProjection` and each complex type to a byte array and subsequently insert the produced byte array into the target `UnsafeArrayData`. Since this logic could be utilized from more places (e.g. `CreateArray`), it should be encapsulated into `UnsafeArrayWriter` or  `UnsafeArrayData` at first. What do you think?

maropu commented:
Also, `postgresql` has the function `array_cat` for concatenating arrays, so it might be better to make the behaviour the same with the `postgresql` one:
https://www.postgresql.org/docs/10/static/functions-array.html

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +289,152 @@ case class ArrayContains(left: Expression, right: Expression)
+  override def dataType: ArrayType =
+    children
+      .headOption.map(_.dataType.asInstanceOf[ArrayType])
+      .getOrElse(ArrayType.defaultConcreteType.asInstanceOf[ArrayType])
```
Should we allow empty children? I can't think of a use case for now and we should better disallow it first.

mn-mikke replied:
Definitely share your opinion, but I think we should be consistent across the whole Spark SQL API. Functions like `concat` and `concat_ws` accept empty children as well.

HyukjinKwon replied:
Hm .. but then this is `array<null>` when the children are empty. Seems `CreateArray`'s type is `array<string>` in this case.

mn-mikke replied:
Ok, changing to return type `array<string>` when no children are provided. Also I've created the jira ticket [SPARK-23798](https://issues.apache.org/jira/browse/SPARK-23798) since I don't see any reason why it couldn't return a default concrete type in this case. Hope I don't miss anything.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1834,6 +1834,25 @@ def array_contains(col, value):
+    Collection function: Concatenates multiple arrays into one.
+
+    :param cols: list of column names (string) or list of :class:`Column` expressions that have
+        the same data type.
```
Shall we note `cols` are expected to be array type?

Commit:
[SPARK-23736][SQL] Code style fixes.

mn-mikke commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -699,3 +699,88 @@ abstract class TernaryExpression extends Expression {
+   * override this.
+   */
+  override def eval(input: InternalRow): Any =
+  {
```
Think I fixed all style differences.

HyukjinKwon replied:
Seems the style fix is missed here.

Commit:
[SPARK-23736][SQL] Improving the description of the ConcatArrays expression.

Commit:
[SPARK-23736][SQL] Merging concat and concat_arrays into one function.

mn-mikke commented:
Merged concat and concat_arrays functions into one via an unresolved expression and subsequent resolution. Do you have any objections to this approach?

SparkQA commented:
**[Test build #88596 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88596/testReport)** for PR 20858 at commit [`bb46c3d`](https://github.com/apache/spark/commit/bb46c3d3d3e18a9e05ddb6fe6efda3c25c2711a4).

Commit:
[SPARK-23736][SQL] Adding new line at the end of the unresolved.scala file.

mn-mikke commented:
retest please

SparkQA commented:
**[Test build #88598 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88598/testReport)** for PR 20858 at commit [`11205af`](https://github.com/apache/spark/commit/11205af476143c188106bb9d42e45fb564d77c14).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88598/
Test FAILed.

Commit:
[SPARK-23736][SQL] Fixing failing unit test from DDLSuite.

mn-mikke commented:
retest please

SparkQA commented:
**[Test build #88605 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88605/testReport)** for PR 20858 at commit [`753499d`](https://github.com/apache/spark/commit/753499d1784bdaf7c96f67dfbc3d0ff5c1e955a9).

SparkQA commented:
**[Test build #88605 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88605/testReport)** for PR 20858 at commit [`753499d`](https://github.com/apache/spark/commit/753499d1784bdaf7c96f67dfbc3d0ff5c1e955a9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88605/
Test FAILed.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1834,6 +1819,25 @@ def array_contains(col, value):
+    [Row(arr=[1, 2, 3, 4, 5]), Row(arr=None)]
+    """
+    sc = SparkContext._active_spark_context
+    return Column(sc._jvm.functions.concat(_to_seq(sc, cols, _to_java_column)))
```
Why did we move this down .. ?

mn-mikke replied:
The whole file is divide into sections according to groups of functions. Based on @gatorsmile's suggestion, the concat function should be categorized as a collection function. So I moved the function to comply with the file structure.

mn-mikke commented:
It seems that we experienced the same problem with failing "RateSourceV2Suite.basic microbatch execution" test reported [here](https://issues.apache.org/jira/browse/SPARK-23504)

Commit:
[SPARK-23736][SQL] Changing method styling according to the standards.

Commit:
[SPARK-23736][SQL] Changing data type to ArrayType(StringType) for the case when no children are provided.

SparkQA commented:
**[Test build #88627 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88627/testReport)** for PR 20858 at commit [`fd84bee`](https://github.com/apache/spark/commit/fd84bee602b41ad463dbee5acd41e3d34581e7ac).

SparkQA commented:
**[Test build #88626 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88626/testReport)** for PR 20858 at commit [`2efdd77`](https://github.com/apache/spark/commit/2efdd771d3d4cc738060d225f621179a44259ebc).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88626/
Test FAILed.

SparkQA commented:
**[Test build #88627 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88627/testReport)** for PR 20858 at commit [`fd84bee`](https://github.com/apache/spark/commit/fd84bee602b41ad463dbee5acd41e3d34581e7ac).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88627/
Test FAILed.

Commit:
[SPARK-23736][SQL] Fixing a SparkR unit test by filtering out UnresolvedConcat from dataset of functions.

SparkQA commented:
**[Test build #88638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88638/testReport)** for PR 20858 at commit [`116f91f`](https://github.com/apache/spark/commit/116f91f5490fb54b4b08e785d881c310e67a2c99).

SparkQA commented:
**[Test build #88638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88638/testReport)** for PR 20858 at commit [`116f91f`](https://github.com/apache/spark/commit/116f91f5490fb54b4b08e785d881c310e67a2c99).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88638/
Test FAILed.

mn-mikke commented:
retest please

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #88661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88661/testReport)** for PR 20858 at commit [`116f91f`](https://github.com/apache/spark/commit/116f91f5490fb54b4b08e785d881c310e67a2c99).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88661/
Test FAILed.

mn-mikke commented:
retest this please

mn-mikke commented:
Can't reproduce any of the failing unit tests above and seem to be unrelated...

maropu commented:
yea, it seems right and I also hit the same failure in my pr.

mn-mikke commented:
The PR [#20688](https://github.com/apache/spark/pull/20688) removes one of the failing unit test from repo. Maybe we could solve the problem by rebasing to current master.

Commit:
[SPARK-23736][SQL] Merging the current master into the feature branch.

SparkQA commented:
**[Test build #88667 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88667/testReport)** for PR 20858 at commit [`e199ac5`](https://github.com/apache/spark/commit/e199ac5bb80349e00b23115793d683a626b005ae).

SparkQA commented:
**[Test build #88667 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88667/testReport)** for PR 20858 at commit [`e199ac5`](https://github.com/apache/spark/commit/e199ac5bb80349e00b23115793d683a626b005ae).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88667/
Test FAILed.

mn-mikke commented:
@tdas Any idea how to fix failing the `RateSourceSuite.overflow`? It seems that more PRs are facing the same problem.

Commit:
[SPARK-23736][SQL] Merging the current master to the feature branch.

SparkQA commented:
**[Test build #88699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88699/testReport)** for PR 20858 at commit [`067c2db`](https://github.com/apache/spark/commit/067c2dbd01b4d44723d432e1f4b99cb926a33315).

SparkQA commented:
**[Test build #88699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88699/testReport)** for PR 20858 at commit [`067c2db`](https://github.com/apache/spark/commit/067c2dbd01b4d44723d432e1f4b99cb926a33315).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88699/
Test PASSed.

mn-mikke commented:
Any other comments?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +289,166 @@ case class ArrayContains(left: Expression, right: Expression)
+ * with concrete concate expressions.
+ */
+object ResolveConcat
+{
```
nit: `object ResolveConcat {`

maropu commented:
Can you add type coercion tests in `sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/concat.sql`?
e.g., the case to concat different array types like concat(array[int], array[long])

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -408,6 +407,7 @@ object FunctionRegistry {
     expression[MapValues]("map_values"),
     expression[Size]("size"),
     expression[SortArray]("sort_array"),
+    expression[UnresolvedConcat]("concat"),
```
I thinks we should not put unresolved expr here.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -700,3 +700,88 @@ abstract class TernaryExpression extends Expression {
+/**
+ * The trait covers logic for performing null safe evaluation and code generation.
+ */
+trait NullSafeEvaluation extends Expression
```
nit: `trait NullSafeEvaluation extends Expression {`

maropu commented:
Also, can you put the gen'd code of concat arrays here for reviews?

mn-mikke commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -3046,6 +3036,16 @@ object functions {
+   * @since 1.5.0
+   */
+  @scala.annotation.varargs
+  def concat(exprs: Column*): Column = withExpr { UnresolvedConcat(exprs.map(_.expr)) }
```
@gatorsmile, @HyukjinKwon What is your view on this?

Commit:
[SPARK-23736][SQL] Merging string concat and array concat into one expression.

Commit:
[SPARK-23736][SQL] Adding more test cases

SparkQA commented:
**[Test build #89010 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89010/testReport)** for PR 20858 at commit [`8abd1a8`](https://github.com/apache/spark/commit/8abd1a8b92eee5b83c13a1969dcbfca7e6cb6a06).

SparkQA commented:
**[Test build #89010 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89010/testReport)** for PR 20858 at commit [`8abd1a8`](https://github.com/apache/spark/commit/8abd1a8b92eee5b83c13a1969dcbfca7e6cb6a06).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89010/
Test PASSed.

mn-mikke commented:
@maropu I've modified the solution according to your comments:
- Removed UnresolvedConcat and merged string and array concatenation into one expression class.
- Implemented type coercion for concatenation of arrays and added tests for it
- Added codegen examples into the description

Please take a look...

Commit:
[SPARK-23736][SQL] Optimizing null elements protection.

SparkQA commented:
**[Test build #89018 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89018/testReport)** for PR 20858 at commit [`367ee22`](https://github.com/apache/spark/commit/367ee2241901225e7451d7280611cecf23be82f1).

SparkQA commented:
**[Test build #89018 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89018/testReport)** for PR 20858 at commit [`367ee22`](https://github.com/apache/spark/commit/367ee2241901225e7451d7280611cecf23be82f1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89018/
Test PASSed.

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,191 @@ case class ArrayContains(left: Expression, right: Expression)
+         |  $counter++;
+         |}
+        """.stripMargin
+    }.mkString("\n")
```
To use `mkString` may lead to a compilation error due to 64KB bytecode limitation. Would it be possible to use `CodegenContext.splitExpressions()`?

gatorsmile commented:
cc @ueshin

Commit:
[SPARK-23736][SQL] Protection against the length limit of Java functions

SparkQA commented:
**[Test build #89286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89286/testReport)** for PR 20858 at commit [`6bb33e6`](https://github.com/apache/spark/commit/6bb33e6eeb2028d64b4258413523e02b08263059).

Commit:
Merge remote-tracking branch 'spark/master' into feature/array-api-concat_arrays-to-master

SparkQA commented:
**[Test build #89299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89299/testReport)** for PR 20858 at commit [`57b250c`](https://github.com/apache/spark/commit/57b250c4641f7a6daebb22a645d014e0bcf90ea4).

SparkQA commented:
**[Test build #89299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89299/testReport)** for PR 20858 at commit [`57b250c`](https://github.com/apache/spark/commit/57b250c4641f7a6daebb22a645d014e0bcf90ea4).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89299/
Test FAILed.

Commit:
[SPARK-23736][SQL] Adding test for the limit of Java function size.

SparkQA commented:
**[Test build #89302 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89302/testReport)** for PR 20858 at commit [`944e0c9`](https://github.com/apache/spark/commit/944e0c9f5312eab9581c961704e875d5bd629878).

SparkQA commented:
**[Test build #89302 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89302/testReport)** for PR 20858 at commit [`944e0c9`](https://github.com/apache/spark/commit/944e0c9f5312eab9581c961704e875d5bd629878).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89302/
Test FAILed.

Commit:
[SPARK-23736][SQL] Adding more tests

SparkQA commented:
**[Test build #89323 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89323/testReport)** for PR 20858 at commit [`7f5124b`](https://github.com/apache/spark/commit/7f5124ba8752387b3e1d6c0922b551a2cba98356).

SparkQA commented:
**[Test build #89323 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89323/testReport)** for PR 20858 at commit [`7f5124b`](https://github.com/apache/spark/commit/7f5124ba8752387b3e1d6c0922b551a2cba98356).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89323/
Test FAILed.

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+        ("UnsafeArrayData", arrayData),
+        ("int[]", counter)))
+
+    s"""new Object() {
```
nit:
```
s"""
   |new Object() {
...
```

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+      if (inputs.contains(null)) {
+        null
+      } else {
+        val elements = inputs.flatMap(_.asInstanceOf[ArrayData].toObjectArray(elementType))
```
Can we always allocate an array? I think that the total array element size may be overflow in some cases.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+      arguments = Seq(
+        (s"${javaType}[]", "args"),
+        ("UnsafeArrayData", arrayData),
+        ("int[]", counter)))
```
I guess we can simply use for-loop here?

```java
for (int $idx = 0; $idx < ${children.length}; $idx++) {
  for (int z = 0; z < args[$idx].numElements(); z++) {
    ...
  }
}
```

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+        |int[] $tempVariableName = new int[]{0};
+        |$assignmentSection
+        |final int $numElementsConstant = $tempVariableName[0];
+      """.stripMargin,
```
I guess we can simply use for-loop here?

```java
int $tempVariableName = 0;
for (int $idx = 0; $idx < ${children.length}; $idx++) {
  $tempVariableName += args[$idx].numElements();
}
final int $numElementsConstant = $tempVariableName;
```

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+         |boolean[] $isNullVariable = new boolean[]{false};
+         |$assignmentSection;
+         |if ($isNullVariable[0]) return null;
+       """.stripMargin
```
I guess we can simply use for-loop here?

```java
for (int $idx = 0; $idx < ${children.length}; $idx++) {
  if (args[$idx] == null) {
    return null;
  }
}
```

We can return as soon as we found `null` in this case.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+    val assignmentSection = ctx.splitExpressions(
+      expressions = assignments,
+      funcName = "complexArrayConcat",
+      arguments = Seq((s"${javaType}[]", "args"), ("Object[]", arrayData), ("int[]", counter)))
```
I guess we can simply use for-loop here?

```java
for (int $idx = 0; $idx < ${children.length}; $idx++) {
  for (int z = 0; z < args[$idx].numElements(); z++) {
    ...
  }
}
```

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+    val assignments = (0 until children.length).map { idx =>
+      s"""
+         |for (int z = 0; z < args[$idx].numElements(); z++) {
+         |  $arrayData[$counter[0]] = ${CodeGenerator.getValue(s"args[$idx]", elementType, "z")};
```
We need to check null?

mn-mikke replied:
Here we operate only with non-primitive types where null is treated as a regular value so the null check shouldn't be necessary.
The added tests should cover this scenario.

Commit:
[SPARK-23736][SQL] Checks of max array size + Rewriting codegen using for loops.

Commit:
[SPARK-23736][SQL] Merging current master into the feature branch.

SparkQA commented:
**[Test build #89402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89402/testReport)** for PR 20858 at commit [`600ae89`](https://github.com/apache/spark/commit/600ae891006b8edda6f817f494e576b65886db47).

SparkQA commented:
**[Test build #89402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89402/testReport)** for PR 20858 at commit [`600ae89`](https://github.com/apache/spark/commit/600ae891006b8edda6f817f494e576b65886db47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
[SPARK-23736][SQL] Fixing exception messages

SparkQA commented:
**[Test build #89456 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89456/testReport)** for PR 20858 at commit [`f2a67e8`](https://github.com/apache/spark/commit/f2a67e82880896bf7c09d3067f7d1699c43d2505).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89456/
Test PASSed.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -353,3 +356,218 @@ case class ArrayMax(child: Expression) extends UnaryExpression with ImplicitCast
+
+  override def dataType: DataType = children.map(_.dataType).headOption.getOrElse(StringType)
+
+  lazy val javaType: String = CodeGenerator.javaType(dataType)
```
We can move this into `doGenCode()` method.

mn-mikke replied:
Good point! But I think it would be better to reuse `javaType` also in `genCodeForPrimitiveArrays` and `genCodeForNonPrimitiveArrays`.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -353,3 +356,218 @@ case class ArrayMax(child: Expression) extends UnaryExpression with ImplicitCast
+    ev.copy(s"""
+      $initCode
+      $codes
+      ${javaType} ${ev.value} = $concatenator.concat($args);
```
nit: `$javaType`

ueshin commented:
LGTM except for nits.

Commit:
[SPARK-23736][SQL] Small refactoring

Commit:
[SPARK-23736][SQL] Merging current master to the feature branch

SparkQA commented:
**[Test build #89504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89504/testReport)** for PR 20858 at commit [`5a4cc8c`](https://github.com/apache/spark/commit/5a4cc8c3d7db2d635902ab6f30394c4246ffe768).

SparkQA commented:
**[Test build #89495 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89495/testReport)** for PR 20858 at commit [`8a125d9`](https://github.com/apache/spark/commit/8a125d9f536269f2aa2f659e9acc99c31e7854e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89495/
Test PASSed.

SparkQA commented:
**[Test build #89504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89504/testReport)** for PR 20858 at commit [`5a4cc8c`](https://github.com/apache/spark/commit/5a4cc8c3d7db2d635902ab6f30394c4246ffe768).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class HasCollectSubModels(Params):`
  * `class Summarizer(object):`
  * `class SummaryBuilder(JavaWrapper):`
  * `class CrossValidator(Estimator, ValidatorParams, HasParallelism, HasCollectSubModels,`
  * `class TrainValidationSplit(Estimator, ValidatorParams, HasParallelism, HasCollectSubModels,`
  * `case class Reverse(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `case class ArrayMin(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `class ArrayDataIndexedSeq[T](arrayData: ArrayData, dataType: DataType) extends IndexedSeq[T] `
  * `abstract class MemoryStreamBase[A : Encoder](sqlContext: SQLContext) extends BaseStreamingSource `
  * `class ContinuousMemoryStream[A : Encoder](id: Int, sqlContext: SQLContext)`
  * `  case class GetRecord(offset: ContinuousMemoryStreamPartitionOffset)`
  * `class ContinuousMemoryStreamDataReaderFactory(`
  * `class ContinuousMemoryStreamDataReader(`
  * `case class ContinuousMemoryStreamOffset(partitionNums: Map[Int, Int])`
  * `case class ContinuousMemoryStreamPartitionOffset(partition: Int, numProcessed: Int)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89504/
Test PASSed.

Commit:
[SPARK-23736][SQL] Merging current master to the feature branch.

SparkQA commented:
**[Test build #89560 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89560/testReport)** for PR 20858 at commit [`f7bdcf7`](https://github.com/apache/spark/commit/f7bdcf789a460ee987961232b116ba5c65cc3cf9).

SparkQA commented:
**[Test build #89560 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89560/testReport)** for PR 20858 at commit [`f7bdcf7`](https://github.com/apache/spark/commit/f7bdcf789a460ee987961232b116ba5c65cc3cf9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ArrayPosition(left: Expression, right: Expression)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
[SPARK-23736][SQL] Merging current master to the feature branch.

SparkQA commented:
**[Test build #89573 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89573/testReport)** for PR 20858 at commit [`36d5d25`](https://github.com/apache/spark/commit/36d5d25a69d5a6134141038c01ada1f5c516b07c).

SparkQA commented:
**[Test build #89573 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89573/testReport)** for PR 20858 at commit [`36d5d25`](https://github.com/apache/spark/commit/36d5d25a69d5a6134141038c01ada1f5c516b07c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ElementAt(left: Expression, right: Expression) extends GetMapValueUtil `
  * `abstract class GetMapValueUtil extends BinaryExpression with ImplicitCastInputTypes `
  * `case class GetMapValue(child: Expression, key: Expression)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89573/
Test PASSed.

ueshin commented:
Thanks! merging to master.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -665,3 +667,219 @@ case class ElementAt(left: Expression, right: Expression) extends GetMapValueUti
+
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def eval(input: InternalRow): Any = dataType match {
```
so this pattern match will probably cause significant regression in the interpreted (non-codegen) mode, due to the way scala pattern matching is implemented.

mn-mikke replied:
Thanks! I've created #22471 to call the pattern matching only once.

WDYT about [Reverse](https://github.com/apache/spark/pull/21034/files#diff-9853dcf5ce3d2ac1e94d473197ff5768R240)? It looks like a similar problem.
